### PR TITLE
Database Service

### PR DESCRIPTION
**Pull request by omermadmon**
`DatabaseService` is a singleton class wrapping the application's datastore instance, provides a global point of access to any read/write operations.

`DatabaseService` provides API for:
- Writing simulation metadata by the builder responsible for the simulation construction (which returns the unique id assigned to the simulation, encoded as a string).
- Writing a board state to the DB (by writing entities under the schema of `[simulationId, round, agentId, agentType, rowNum, colNum]` ).
- Reading a board state from the DB (which returns null if provided round exceeds simulation's max number of round, or a JSON string encoding the board state).

In addition, `DatabaseService`  throw informative exceptions whenever an invalid read/write operation occurs (e.g when writing a board state which already exists for a certain `simulationId` and `round`).

### CODE REVIEW

File: src/main/java/com/google/research/bleth/services/DatabaseService.java
Comment (omermadmon) [reply src/main/java/com/google/research/bleth/services/DatabaseService.java:219]: I agree something needs to be changed, though I am not sure if we should change it here or at the Simulation's `run()` method (to be `currentRound < maxNumberOfRounds` as well).

It depends whether we consider the initial state (round zero) as a round, or not.

In my opinion it is more reasonable to count it as a round, otherwise we will have rounds 0 to maxRound recorded at out db - these are maxRound + 1 rounds instead of maxRound rounds (which is a bit weird, I think).

(In other words - I think we want the last round to be the maxRound - 1 round, thus we need to change the while loop condition of the simulation, and leave it as is here).

File: src/main/java/com/google/research/bleth/exceptions/ExceedingRoundException.java
Comment (omermadmon) [reply src/main/java/com/google/research/bleth/exceptions/ExceedingRoundException.java:6]: It is now my turn to quote Alon:

"Prefer to create your own exception classes, rather than inventing new semantics for existing exceptions."

File: src/main/java/com/google/research/bleth/services/DatabaseService.java
Comment (OrpazFishel) [reply src/main/java/com/google/research/bleth/services/DatabaseService.java:219]: I assume that considering the initial state as round 0 will be more comfortable, but it's your call. Just make sure it's consistent.

File: src/main/java/com/google/research/bleth/exceptions/ExceedingRoundException.java
Comment (OrpazFishel) [reply src/main/java/com/google/research/bleth/exceptions/ExceedingRoundException.java:6]: Oh I do remember this comment, I just don't see why this is new semantics. But again - it's your call.

File: src/main/java/com/google/research/bleth/services/DatabaseService.java
Comment (OrpazFishel) [inline src/main/java/com/google/research/bleth/services/DatabaseService.java:177]: Nit: Is there any reason for using Long instead of long?

File: src/main/java/com/google/research/bleth/services/DatabaseService.java
Comment (OrpazFishel) [inline src/main/java/com/google/research/bleth/services/DatabaseService.java:70]: Do all the observers have the same strategy? If so - it's observer**s** and not observer, otherwise we should save this information about each agent.

File: src/test/java/com/google/research/bleth/services/DatabaseServiceTest.java
Comment (OrpazFishel) [inline src/test/java/com/google/research/bleth/services/DatabaseServiceTest.java:110]: Consider mentioning that you should get this Json when you read from db - the name only mentions writing to it.

File: src/test/java/com/google/research/bleth/services/DatabaseServiceTest.java
Comment (OrpazFishel) [inline src/test/java/com/google/research/bleth/services/DatabaseServiceTest.java:221]: Why don't you use here ZERO_ROUND?

File: src/main/java/com/google/research/bleth/services/DatabaseService.java
Comment (omermadmon) [reply src/main/java/com/google/research/bleth/services/DatabaseService.java:177]: Yes, the `intValue()` method.

File: src/test/java/com/google/research/bleth/services/DatabaseServiceTest.java
Comment (omermadmon) [reply src/test/java/com/google/research/bleth/services/DatabaseServiceTest.java:110]: writeNonEmptyRealBoard**ThenRead**RealBoard_shouldGetExpectedJson

File: src/main/java/com/google/research/bleth/services/DatabaseService.java
Comment (omermadmon) [reply src/main/java/com/google/research/bleth/services/DatabaseService.java:217]: Resolved at `omermadmon-simulation` branch :)

File: src/main/java/com/google/research/bleth/services/DatabaseService.java
Comment (schainmariano) [inline src/main/java/com/google/research/bleth/services/DatabaseService.java:39]: purpose of this?

File: src/main/java/com/google/research/bleth/services/DatabaseService.java
Comment (schainmariano) [inline src/main/java/com/google/research/bleth/services/DatabaseService.java:67]: the database service seems to be aware of internal structure of the simulation object (e.g., why should it care about the movement strategy?). this creates a dependency in the wrong direction.

File: src/main/java/com/google/research/bleth/services/DatabaseService.java
Comment (schainmariano) [inline src/main/java/com/google/research/bleth/services/DatabaseService.java:160]: can this call fail?

File: src/main/java/com/google/research/bleth/services/DatabaseService.java
Comment (schainmariano) [inline src/main/java/com/google/research/bleth/services/DatabaseService.java:166]: this highly couples the database utility with the simulation implementation/interface

File: src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java
Comment (schainmariano) [inline src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java:11]: if this is an abstract class then you should not implement pure abstract methods (that is, do not "return 1")

**Review by schainmariano — CHANGES_REQUESTED** [review]
my main concern is the dependency of the service on the internals of the client

File: src/main/java/com/google/research/bleth/services/DatabaseService.java
Comment (omermadmon) [reply src/main/java/com/google/research/bleth/services/DatabaseService.java:67]: The idea here was that the `builder` will pass himself to the `writeSimulationMetadata()` method inside its `build()` method. Something like:

```java
public TracingSimulation build() {
    this.id = DatabaseService.getInstance().writeSimulationMetadata(this);
    return new TracingSimulation(this);
}
```

Alternatively we can modify `writeSimulationMetadata()` to recieve multiple arguments instead of a builder, and then our `build()` method will look like:

```java
public TracingSimulation build() {
    this.id = DatabaseService.getInstance().writeSimulationMetadata(this.rowNum, this.colNum, /** lots of other attributes */);
    return new TracingSimulation(this);
}
```

But this smells bad as well. I guess what we're talking about is having some other object in the middle, which specifies both the properties names and the values to be written to the db. Let's say we're using some kind of a HashMap:

```java
public TracingSimulation build() {
    HashMap<String, Double> metadataEntity = new HashMap<>();
    metadataEntity.put("rowNum", this.rowNum);
    metadataEntity.put("colNum", this.colNum);
    /** lots of other attributes */
    this.id = DatabaseService.getInstance().write(metadataEntity);
    return new TracingSimulation(this);
}
```

And this might be better, because now the `write()` method can be similar for all entities we're writing. But someone must create this HashMap (or whatever we will work with) everytime, including specifing the properties names. 

Did you think of something else? @schainmariano

Comment (omermadmon) [reply src/main/java/com/google/research/bleth/services/DatabaseService.java:67]: Also, the reason we're writing the Simulation's strategies as part of the metadata is to able to provide answers for questions like "what was the average result in all simulations where the beacons moved **randomally** and the observers where **static**?"

File: src/main/java/com/google/research/bleth/services/DatabaseService.java
Comment (omermadmon) [reply src/main/java/com/google/research/bleth/services/DatabaseService.java:166]: Here the database utility is dependent on the datastore Simulation entity structure (i.e its schema), not the simulation implementation/interface. How can we avoid this? when reading an entity from datastore (or any other relational/non-relational db) we must specify the entity's properties names.

@schainmariano

File: src/main/java/com/google/research/bleth/services/DatabaseService.java
Comment (omermadmon) [reply src/main/java/com/google/research/bleth/services/DatabaseService.java:160]: - `asIterable()` throws an exception if the query being executed is associated with a `Transaction` that is not active:
https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/PreparedQuery.html#asIterable--
This doesn't seem to be the case here since we're not using Datastore Transactions.
- `toJsonTable()` throws an exception if the simulation associated with the specified `simulationId` does not exist in datastore. But if that is the case, an exception will be thrown even before that, right when there is a call for `isRoundExistsInSimulation(simulationId, round)` (line 141).

These are all unchecked exception, therefore there is no use of the `throws` keyword in the method's signature. I will add some documentation to explain this behaviour.

File: src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java
Comment (omermadmon) [reply src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java:11]: This `AbstractSimulation` is a placeholder, and it exists only to provide a mock simulation in tests. This implementation will be overridden as soon as PR #6 is merged to master.

**Comment by omermadmon** [comment]
We are aiming for a DB service which is more flexible and agnostic, hence we can just use the Datastore API without writing a wrapper class.

In addition, I have added the following classes: 
- `BoardState` -  represents a static `Board` associated with a specific `simulationId` and `round`.
Provides API for writing to db, (statically) reading from db and JSON serialization.
- `SimulationMetadata` - stores all metadata associated with a simulation.
Provides API for writing to db and (statically) reading from db.
- `Schema` - stores an provides a single access point to all Datastore entities kinds and schemas.

This way our database service does not depend on its clients, and also only the `BoardState` and `SimulationMetadata` classes depend on the database service.
i.e: if we decide to use Cloud SQL instead of Datastore, we know we will have to change only those classes.

File: src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java
Comment (OrpazFishel) [inline src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java:33]: Nit: Consider re-write the sentence.

File: src/main/java/com/google/research/bleth/simulator/BoardState.java
Comment (OrpazFishel) [inline src/main/java/com/google/research/bleth/simulator/BoardState.java:42]: Nit: Indentation.

File: src/main/java/com/google/research/bleth/simulator/BoardState.java
Comment (OrpazFishel) [inline src/main/java/com/google/research/bleth/simulator/BoardState.java:68]: Nit: Did you mean "represents an agent"?

File: src/main/java/com/google/research/bleth/simulator/BoardState.java
Comment (OrpazFishel) [inline src/main/java/com/google/research/bleth/simulator/BoardState.java:74]: Nit:  Did you mean "an exception"?

File: src/main/java/com/google/research/bleth/simulator/BoardStateFactory.java
Comment (OrpazFishel) [inline src/main/java/com/google/research/bleth/simulator/BoardStateFactory.java:42]: Nit: Indentation.

File: src/main/java/com/google/research/bleth/simulator/BoardStateFactory.java
Comment (OrpazFishel) [inline src/main/java/com/google/research/bleth/simulator/BoardStateFactory.java:69]: Make sure the board isn't null.

File: src/main/java/com/google/research/bleth/simulator/SimulationMetadata.java
Comment (OrpazFishel) [inline src/main/java/com/google/research/bleth/simulator/SimulationMetadata.java:49]: Nit: Consider re-write the sentence.

File: src/main/java/com/google/research/bleth/simulator/SimulationMetadata.java
Comment (OrpazFishel) [inline src/main/java/com/google/research/bleth/simulator/SimulationMetadata.java:70]: The parentheses aren't necessary.

File: src/main/java/com/google/research/bleth/simulator/SimulationMetadata.java
Comment (OrpazFishel) [inline src/main/java/com/google/research/bleth/simulator/SimulationMetadata.java:92]: Consider renaming to isRoundExistsInSimulation.

**Review by schainmariano — APPROVED** [review]

**Pull request merged**
